core/class.go

```diff
@@ -46,8 +57,270 @@ func (c *Cairo0Class) Version() uint64 {
 	return 0
 }
 
+//nolint:funlen
 func (c *Cairo0Class) Hash() (*felt.Felt, error) {
-	return cairo0ClassHash(c)
+	definition, err := makeDeprecatedVMClass(c)
+	if err != nil {
+		return nil, err
+	}
+
+	var program Program
+	err = json.Unmarshal(definition.Program, &program)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use goroutines to parallelize hash computations
+	var wg sync.WaitGroup
+	var externalEntryPointHash, l1HandlerEntryPointHash, constructorEntryPointHash, builtInsHash, dataHash *felt.Felt
+	var hintedClassHash *felt.Felt
+	var hintedClassHashErr error
```

Comment:
I think we can improve this logic by:
1. replace `sync.WaitGroup` with `conc.NewWaitGroup()`
2. move common logic for EntryPoints in lambda and resuse for `External`, `L1Handler` and `Constructor`
3. as a consequence remove all digest variables below

i.e.:
```
	var externalEntryPointHash, l1HandlerEntryPointHash, constructorEntryPointHash, builtInsHash, dataHash *felt.Felt
	var hintedClassHash *felt.Felt
	var hintedClassHashErr error

	entryPointsDigest := func(entryPoints []starknet.EntryPoint) *felt.Felt {
		var epDigest crypto.PedersenDigest
		for _, ep := range entryPoints {
			epDigest.Update(ep.Selector, ep.Offset)
		}
		return epDigest.Finish()
	}
	wg := conc.NewWaitGroup()
	wg.Go(func() {
		externalEntryPointHash = entryPointsDigest(definition.EntryPoints.External)
	})
	wg.Go(func() {
		l1HandlerEntryPointHash = entryPointsDigest(definition.EntryPoints.L1Handler)
	})
	wg.Go(func() {
		constructorEntryPointHash = entryPointsDigest(definition.EntryPoints.Constructor)
	})
	wg.Go(func() {
		var builtInsDigest crypto.PedersenDigest
		for _, builtIn := range program.Builtins {
			builtInHex := hex.EncodeToString([]byte(builtIn))
			builtInFelt, err := new(felt.Felt).SetString("0x" + builtInHex)
			if err != nil {
				return
			}
			builtInsDigest.Update(builtInFelt)
		}
		builtInsHash = builtInsDigest.Finish()
	})
	wg.Go(func() {
		hintedClassHash, hintedClassHashErr = computeHintedClassHash(definition.Abi, definition.Program)
	})
	wg.Go(func() {
		var dataDigest crypto.PedersenDigest
		for _, data := range program.Data {
			dataFelt, err := new(felt.Felt).SetString(data)
			if err != nil {
				return
			}
			dataDigest.Update(dataFelt)
		}
		dataHash = dataDigest.Finish()
	})
	wg.Wait()
```

Reply:
and btw. are you sure that `.SetString("0x" + builtInHex)` is necessery?

Reply:
Yeah, `hex.EncodeToString` outputs the hexadecimal representation without `0x` prefix. `felt.SetString` requires the parsed string to have `0x` prefix. You can check `felt_test.go` and verify the usage.

Here's a quick code snippet that you can verify hex encoding output in [go playground](https://go.dev/play/):
```
package main

import (
	"encoding/hex"
	"fmt"
)

func main() {
	// Sample data to encode
	data := []byte("Hello, Hex Encoding!")

	// Encode the data to hexadecimal string
	hexString := hex.EncodeToString(data)

	// Print the result
	fmt.Printf("Original: %s\n", data)
	fmt.Printf("Hex encoded: %s\n", hexString)
}
```